### ISSUE TITLE
How can we enable continuous learning with the LLM model ? 

### ISSUE THREAD

**Issue by TapendraBaduwal** [huggingface/peft#1280]
### Feature request

If we merge adapters then this work for continuous learning ?
I expect the new train model to provide output for previously trained data as well.

```python
model_path = "model/path" 

# Reload model in FP16 and merge it with LoRA weights
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch.float16,
    device_map=device_map,
)
model = PeftModel.from_pretrained(base_model, new_model)
model = model.merge_and_unload()

# Reload tokenizer to save it
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"

# Save the merged model
model.save_pretrained(model_path)
tokenizer.save_pretrained(model_path)
```

### Motivation

continuous learning

### Your contribution

continuous learning

**Comment by BenjaminBossan** [comment]
> If we merge adapters then this work for continuous learning ?
> I expect the new train model to provide output for previously trained data as well.

If I understand you correctly, you're asking if you can fine-tune a model on task A, merge the adapter, then add another adapter and fine-tune on task B without the model forgetting about task A. This is certainly possible. During training on task B, you should monitor the performance on task A too, to ensure that there is no catastrophic forgetting. Since you only train the B-adapter, it should be less likely to encounter this, though. What could happen is that B interferes with A. Experiment with different hyper-parameter settings or try to tweak the dataset if you see this issue.

**Comment by TapendraBaduwal** [comment]
@BenjaminBossan During training on task B, you should monitor the performance on task A too, to ensure that there is no catastrophic forgetting.
This means while training on task B we have to mix some data points from task A too ?

**Comment by BenjaminBossan** [comment]
> This means while training on task B we have to mix some data points from task A too ?

No, for training, you can stick with data from task B. What I meant is ideally, every few training steps, when you evaluate the model, you should also evaluate it on task A. That way, you can see early signs of performance degradation on task A. Of course, you can evaluate on task A after you finished training on task B, but when the scores are bad for A, you lost a lot of time. That's why I recommend checking regularly.

> And then we need to merge model on task A adapter with model on task B ?

Not necessarily. You have merged the LoRA for task A and the LoRA for task B would be unmerged, but that doesn't really matter for the quality of the inference. You might want to merge LoRA for task B to increase inference speed though.

**Comment by TapendraBaduwal** [comment]
@BenjaminBossan 
Since I am testing on 1000 data points from Task A where the response= xyz, Here is my Task B data point with a response= abc. If I train model1 for Task A and fine-tune model2 on model1, my model2 is not able to respond with 'xyz which is response of model1 for taskA.

**Comment by BenjaminBossan** [comment]
Hmm, if I understand you correctly, your task A and task B have conflicting targets for the same input. In that case, what would you expect the right answer to be? If task B is correct and task A is incorrect, why train on task A in the first place? If task B is somehow an improved version of task A, then probably you don't need to check task A when training on task B.

**Comment by TapendraBaduwal** [comment]
TaskA dataset as:
{"instruction": "Animal which makes a 'mew' sound?",
  "context": "Describe the characteristics and traits of a cat.",
  "response": "cat" }
.............................

TaskB dataset as:
{"instruction": "Animal which makes a 'bhaw' sound?",
 "context": "Describe the characteristics and traits of a dog.",
  "response": "dog"
 }

At first i trained an LLM model on Task A, and then I fine-tuned the new model by using the previously model from Task A. However, my newly trained model is unable to produce the response 'cat' when i asked about a cat.
In future if any new data will be there like 'human' as response i want to finetune model on previously train model without taking old dataset of cat and dog.
How can i do continuous learning like that ?

**Comment by BenjaminBossan** [comment]
This is a very tough question to answer in the abstract. When you get new data with a new target, I would suggest training the base model with a completely new adapter on all data, instead of trying to train on top of the adapter of task A. If you don't have the resources for that, you probably need to experiment with the hyper-parameters (e.g. learning rate) to ensure that training on a new task does not lead the model to forget about the old task.
